torch/csrc/jit/python/pybind_utils.h

```diff
@@ -128,9 +130,10 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
       if (value.isPyObject()) {
         pybind11::gil_scoped_acquire gil;
         py::object obj = torch::jit::toPyObject(value);
-        // FIXME Should we support more types than just tensor lists?
-        auto new_value = torch::jit::toIValue(
-            obj, c10::ListType::create(c10::TensorType::get()));
+        // FIXME This could fail. As a fallback we could try to pickle the
+        // object, since the pickler might support broader types and it is able
+        // to extract the tensors from the object as a vector.
+        auto new_value = torch::jit::toTypeInferredIValue(obj);
```

Comment:
My understanding is that the type can only be inferred when it's "regular", i.e., a mixed-type list won't work. Are there ways to handle such cases? Would pickling work?

More in general, I'd like to understand more about the PyObject type. Are all objects coming from Python represented as PyObjects, or only the ones for which no better type is found? Do we always need to hold the GIL when manipulating such types?